### ISSUE TITLE
FormRequest: inner objects get nullified during validation // Проблема со сложной структурой параметров.

### ISSUE THREAD

**Issue by s-chizhik** [upgate/laravel-jsonrpc#28]
4fdace42eccf39f07abc72c8aa67c9d07bf4bcaa (v0.7.0)

Здравствйте. Столкнулся с проблемой. Если использовать правила валидации через дочерние классы 
 `Upgate\LaravelJsonRpc\Server\FormRequest` для внутренних структур типа `field.*.field` то эти поля после валидации примут значение `null`.

Пример:

```PHP
class TestRequest extends \Upgate\LaravelJsonRpc\Server\FormRequest
{
    public function rules(): array
    {
        return [
            'simpleParam1' => 'string',
            'simpleParam2' => 'integer',
            'complexParamArray' => 'sometimes|array',
            'complexParamArray.*.a' => 'integer',
            'complexParamArray.*.b' => 'string',
        ];
    }
}
```

Запрос:
```
{
  "id": 1597253370,
  "method": "test.test",
  "jsonrpc": "2.0",
  "params": {
    "simpleParam1": "string",
    "simpleParam2": 42,
    "complexParamArray": [
      {
        "a": 1,
        "b": 2
      },
      {
        "a": 3,
        "b": 4
      }
    ]
  }
}
```

Результат `dd($request->all())`
```
array:4 [
  "simpleParam1" => "string"
  "simpleParam2" => 42
  "complexParamArray" => array:2 [
    0 => {#1218
      +"a": null
      +"b": null
    }
    1 => {#1009
      +"a": null
      +"b": null
    }
  ]
]
```

Происходит это из-за этой логики https://github.com/laravel/framework/blob/c7d954003627b46d54d990c354d521384c80edf7/src/Illuminate/Validation/ValidationData.php#L43 в коде Laravel. Она копирует параметры запроса, и не рассчитывает, что внутри могут быть объекты, а поэтому она обнуляет их свойства, а объекты передаются по ссылке.

А здесь https://github.com/upgate/laravel-jsonrpc/blob/master/src/Server/RequestFactory.php#L21 запрос парсится в объект.

Если делать аналогичный JSON запрос `POST`-ом в Laravel используя аналогичные правила валидации — такой проблемы не возникает.

---

Я предполагаю два решения:
- можно попробовать затащить в Laravel PR и попытаться исправить это там
- сделать PR сюда и парсить тело запроса в ассоциативный массив, но я не уверен, насколько пользователи вашей библиотеки используют такие вложенные структуры и как это скажется на них

Я склоняюсь ближе ко второму варианту, потому что Laravel сама так делает: https://github.com/laravel/framework/blob/c7d954003627b46d54d990c354d521384c80edf7/src/Illuminate/Http/Request.php#L354

**Comment by kbaryshnikov** [comment]
Thanks for the report!

Please use English. (For those who don't speak Russian, here's a summary of the issue: inner objects get nullified during validation, and that's how Illuminate\Validation works: it does not expect by-ref values).

I don't like the idea to parse into arrays. First of all, it's not required to use LaravelJsonRpc\Server\FormRequest, it's a convenience feature that does not fit all the use cases. If there's an object, a handler might be expecting an object, and there are cases when it's important to distinguish between objects and arrays. For example, imagine a JSON-RPC 2 call with an inner JSON-RPC 2 call in a method parameter (I'm not saying that's a good idea, it's a terrible design, but it works as an example): a plain method call will get transformed into a malformed batch.

And, most importantly, this is a public open source library, and people might be relying on current behavior.

As for a Laravel PR, I believe it won't be accepted: while Laravel tries to look modular, in fact, many components rely on assumptions which are true for Laravel internals but may be false for outer code, and no one cares. This case is not the worst one. You may try of course, but I wouldn't expect anything.

But I agree that the issue exists, and it needs to be fixed. 

The core of the issue here is Illuminate\Validation not cloning data properly. As I want the fix to be backwards compatible, the only way I see is to deep clone $data before passing to ValidationFactory. That's what PR #29 does, could you please review it?

**Comment by s-chizhik** [comment]
At first, sorry for the Russian language. I just saw a few issues on it, but I've mixed it up with another repository :(

Second, thanks a lot for the quick response and PR!

> If there's an object, a handler might be expecting an object, and there are cases when it's important to distinguish between objects and arrays

I agree.

> And, most importantly, this is a public open source library, and people might be relying on current behavior.

Sure, so I decided not to propose  radically change the type of parsing

> As for a Laravel PR, I believe it won't be accepted

Unfortunately, I agree, but if I have a bit time, I'll try to send PR, at least for next version.

> That's what PR #29 does, could you please review it?

Already. Looks good to me.

**Comment by kbaryshnikov** [comment]
> Looks good to me.

Thanks!

Merged, 0.7.1 released.

**Comment by s-chizhik** [comment]
Thanks for your work! :+1:
